### PR TITLE
[macOS] Remove isComposing workaround

### PR DESCRIPTION
**Pull request by knopp**
Fixes https://github.com/flutter/flutter/issues/85328

Framework PR: https://github.com/flutter/flutter/pull/105407

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by justinmc — COMMENTED** [review]
This looks straightforward to me.  Here's my framework PR review: https://github.com/flutter/flutter/pull/105407#pullrequestreview-998681986

If we're going to move forward with this we should test that kPerformIntent is invoked.

**Comment by chinmaygarde** [comment]
cc @justinmc Are you able to review this?

**Comment by knopp** [comment]
Marking WIP: Needs better way to group multiple selectors (i.e. alt+up sends `moveBackward:` and `moveToBeginningOfParagraph:`. Ideally we want to process both in one microtask.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:687]: This doesn't seem to be correct according to the documentation https://developer.apple.com/documentation/appkit/nstextinputclient/1438246-setmarkedtext?language=objc

> replacementRange
The range to replace, computed from the beginning of the marked text.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:653]: > Will this cause some events to reorder, for example, if a key is bound to 3 selectors [moveBackward, insert: a, moveForward], since insert:a will be handled by the input plugin, this essentially becomes [insert: a, moveBackward, moveForward], from the framework's perspective?

This is a good point. I'll test this. 

> Maybe use dispatch_async instead of CFRunLoopPerformBlock? Then you don't have to set the mode?

We can't pump the dispatch queue manually in unit tests. We can however run the run loop in custom mode so that only these blocks are executed.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:687]: Can you be more specific? I set the replacement range as composing range, that way it will be replaced in `TextInputModel::UpdateComposingText`.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:653]: > Will this cause some events to reorder, for example, if a key is bound to 3 selectors [moveBackward, insert: a, moveForward], since insert:a will be handled by the input plugin, this essentially becomes [insert: a, moveBackward, moveForward], from the framework's perspective?

So it will indeed become that. However even if we flush buffered selectors and it will become [moveBackward, insert:a, moveForward] it still won't get expected result, because when `TextInputPlugin` processes `insert:a` the selection hasn't been updated by the framework yet so that first [moveBackward:] will only update selection after text is already inserted.

So I don't think we can handle compound actions that involve inserting text correctly at this point. It may required some rather significant changes to get this working.

**Review by justinmc — COMMENTED** [review]
Just jumping in with a few nits here as I read through this PR.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:637]: There's a `noop:` selector (or maybe `noop`) when a shortcut isn't mapped to anything.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:580]: The `+ `  is part of diff. Actual method starts with `- (void)` :)

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm]: > We can't pump the dispatch queue manually in unit tests. We can however run the run loop in custom mode so that only these blocks are executed.

I thought we could add a dispatch barrier and use the async expectation API to verify the side effects?

Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm]: Something still needs to pump the main dispatch queue. I don't know any other way to do this except running the [CFRunLoop in default mode](https://github.com/opensource-apple/CF/blob/3cc41a76b1491f50813e28a4ec09954ffa359e6f/CFRunLoop.c#L1590). However running the `CFRunLoop` in default mode in tests crashes because of various scheduled `NSView` related tasks from other tests.

Dispatch barrier is unlikely to help here. Doing anything blocking on main thread (`dispatch_sync`, `dispatch_barrier_sync`, `dispatch_barrier_async_and_wait`) will crash with "LIBDISPATCH: dispatch_sync called on queue already owned by current thread". And async barrier will do not anything useful (it still needs the queue to be pumped).

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: It says `replacementRange` is computed from the beginning of the marked text so it sounds like the starting point should be calculated from the current marked text.

Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: `replacementRange` is specified as being computed relative to the beginning of the marked text (as opposed to from the beginning of the entire text).

My memory is a bit fuzzy since it's been a while since I hacked on this, but I believe the range to replace is:

start: markedRange.location + replacementRange.location
end: markedRange.location + replacementRange.location + replacementRange.length

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm]: This test passes (didn't need a barrier since the main queue is not concurrent):
```swift
    func testExample() throws {
      var x = 0
      let expectation = XCTestExpectation()
      DispatchQueue.main.async {
        print("execute")
        x += 1
      }
      DispatchQueue.main.async {
        XCTAssertEqual(x, 1)
        expectation.fulfill()
      }
      print("wait")
      wait(for: [expectation], timeout: 10)
    }
```

Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm]: `xctwaiter.wait` calls `_dispatch_main_queue_drain()`, so the main queue is pumped in a sense.

Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm]: `_dispatch_main_queue_drain()` doesn't seem to be exported from libdispatch. Looking at XCTest it seems that `waitFor` actually [runs the run loop](https://github.com/apple/swift-corelibs-xctest/blob/dd599d04b0d17e5be3de5fd31101e7edda5d2b42/Sources/XCTest/Public/Asynchronous/XCTWaiter.swift#L367), which would have same sideeffect as calling `CFRunLoopRunInMode`with default modes manually, as in it would crash `flutter_desktop_darwin_unittests`. Now why this crash happens is another issue (perhaps we could drain the run loop in each test while the appropriate cocoa objects still exist), but that should be tracked as another issue.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: I don't see how that would work. With character alternates selection 
Screenshot 2022-06-23 at 06 16 04
`- (void)setMarkedText:selectedRange:replacementRange:` is first called without composing being active and with `replacementRange` provided. So at that point _there is no marked range_. This happens when you select first candidate. The offset specified is clearly from the beginning of the string (it replaces the long pressed character).

I looked at what [Chromium does](https://source.chromium.org/chromium/chromium/src/+/main:content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm;drc=e4e66f3618a6a2a2d05d87ab98431045baaf3006;l=1976) and as far as I can tell it also sets the marked range to replacement range (from the beginning of string) - seems same as what this PR does. 

And this is what [webkit does](https://github.com/WebKit/WebKit/blob/bfe3f272a5b8cc261b68249303d357fd424ec572/Source/WebKit/WebProcess/WebPage/WebPage.cpp#L6135). The `replacementEditingRange` argument is the `replacementRange`, and the whole thing is set as selection before updating as text.

I think the documentation might be wrong about this one, at very least incomplete since the `replacementRange` is provided in situation when there is no marked range.

Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: Agreed the documentation is so ambiguous as to be close to useless. When I wrote the initial UIKit equivalent, the Apple APIs there were at least a little cleaner, but docs weren't much better and much of it was figured out purely by manually checking the behaviour :/

Curious -- what's the behaviour is in a case where this is called where there is existing mark text? I'm assuming that would get called when, for example hiragana is converted to a kanji candidate with Japanese input? To test:
1. Type "hello" so you're offset from the zero position of the string
2. switch to Japanese input
3. set hiragana mode from the input dropdown
4. type nihonngo (you should get にほんご)
5. press space bar and select 日本語 (it may convert automatically without the space depending on your input history)

I'm guessing that during the hiragana input, this might get called multiple times as the conversions happen. Are the replacementRange values always relative to 0 in those cases, or just when there's no mark region? Agreed it would be weird to have two behaviours depending on the presence of an existing mark region, but curious. 

Assuming that behaves well with a CJK mark region, then seems reasonable. Thanks for checking.

Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: The `replacementRange` with any kind of CJK composition always seems to be `NSNotFound`. I tried your example and it was `NSNotFound` as well. The only way I can trigger a non-empty `replacementRange` is the long press character alternate selection, which starts with no composition and then switches to composition upon first selection. Even here the `replacementRange` is only used initially (when not compositing), and in subsequent selection changes the `replacementRange` is empty and (now non-empty)`markedRange` is used.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:582]: Bit of topic, but since flutter text doesn't define tab shortcut and thus it's always handled as focus, when I wanted to insert tab I simply used this Action
```dart
class InsertTabAction extends Action<NextFocusIntent> {
  final TextEditingController controller;

  InsertTabAction(this.controller);

  void _sendIntent(Intent intent) {
    final BuildContext? primaryContext = primaryFocus?.context;
    if (primaryContext != null) {
      Actions.invoke(primaryContext, intent);
    }
  }

  @override
  Object? invoke(NextFocusIntent intent) {
    _sendIntent(ReplaceTextIntent(controller.value, '', controller.selection,
        SelectionChangedCause.keyboard));
    _sendIntent(ReplaceTextIntent(controller.value, '\t', controller.selection,
        SelectionChangedCause.keyboard));
    return null;
  }
}
```
and bound it to `NextFocusIntent` above text editor.  It's not the most obvious solution as it requires binding to `NextFocusIntent` instead of using Shortcut and custom intent, but it works even with Mac IME input.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:580]: Right you are :dart:

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:582]: Can you illustrate why using `Shortcut` with `SingleActivator(LogicalKeyboardKey.tab)` and some `InsertTabIntent` wouldn't work?

Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:582]: > Can you illustrate why using `SingleActivator(LogicalKeyboardKey.tab)` and some `InsertTabIntent` in a `Shortcuts` above the text editor wouldn't work?

It wouldn't. That's why I said it requires binding to `NextFocusIntent`. Tab has to be unconditionally passed to the IME (because we have no way of knowing whether the candidates popup is shown), and when we receive `insertTab:` from TextInputContext we send `NextFocusIntent` to current focus node. This is not ideal, but I don't think there's any way currently to synthesize key event and activate shortcuts, is there?

**Comment by cbracken** [comment]
@knopp is this good to land?

**Comment by dkwingsmt** [comment]
Sorry I'm still taking a look, trying to understand the data flow after the change.

**Comment by knopp** [comment]
> @knopp is this good to land?

I'd prefer landing https://github.com/flutter/engine/pull/34250 first and then rebasing this one.

> Sorry I'm still taking a look, trying to understand the data flow after the change.

As far as engine is concerned the only change is that editing events are not forwarded to `TextInputContext` directly when compositing. The framework needs to make sure that events possibly related to compositing (mostly arrows keys/tab/enter/esc) are unhandled. Engine then translates this to selectors and in framework the selectors are translated to intents and sent to active focus node.

**Comment by dkwingsmt** [comment]
After talking with @LongCatIsLooong I think I understand the flow better. Let me phrase it and see if it's correct.

------

After this and https://github.com/flutter/flutter/pull/105407, certain control keys (tab, arrows, etc.) will by rejected by the framework when a text input is being focused (formally, these keys are mapped to `DoNothingAndStopPropagationTextIntent`). 

These events are sent back to the AppKit platform. Now, I assume, is when the platform gets to dispatch them to the IME (emoji picker.) 

If the emoji picker doesn't handle them, the platform dispatches them to `doCommandBySelector` as actions represented as strings. (For this to work `FlutterTextInputPlugin` needs to implement `insertTab` with an empty body.)

These string-actions are sent to Framework's `TextInput` through a new channel, whose callback converts these messages into intents.

Since these string-actions now always dispatch fixed intents, for example, a Tab key always dispatches `NextFocusIntent`, the developer can change their behavior by overriding the intent-to-action map.

------

If so, I have a few questions:

First, the framework change only rejects control keys without any modifiers. For example, it rejects ArrowLeft, but not Alt-ArrowLeft, but Alt-ArrowLeft is still removed from `_macShortcuts`. It seems to me that this should not work. If they do work after the change, does mean these extra default text actions (such as for Alt-ArrowLeft) were not needed?

Second, while I'm still concerned with forcing the developer with override a NextFocusIntent with an InsertTabAction, I can't think of a better solution. However, this requires much more education, since overriding stuff is more confusing. We need a way to tell the developers the intent-to-action map, instead of the activator-to-intent map (which is understandably more intuitive), and the breakage otherwise will only be noticed in the corner case of emoji picker. Moreover, they will have to remember the enumerated intents to override, which basically means they're associating `NextFocusIntent` with pressing a `Tab` key.

I'm wondering if it will be a better choice to create a `MacosSelectorIntent(String content)` and map them to the actions for macOS. This also makes it look less weird to map to an `InsertTabAction`. What do you think?

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:690]: Done.

**Comment by knopp** [comment]
> If so, I have a few questions:
> 
> First, the framework change only rejects control keys without any modifiers. For example, it rejects ArrowLeft, but not Alt-ArrowLeft, but Alt-ArrowLeft is still removed from `_macShortcuts`. It seems to me that this should not work. If they do work after the change, does mean these extra default text actions (such as for Alt-ArrowLeft) were not needed?

It rejects ArrowLeft, because it is mapped to `DirectionalFocusIntent(TraversalDirection.left)` inside `app.dart`. So during text editing it needs to be mapped to `DoNothingAndStopPropagationTextIntent` in order for framework to not handle the event. 

Alt-ArrowLeft on the other hand is not mapped to anything in framework. So framework does not handle the event, the engine will pass it to TextInputContext, which will produce the `moveWordLeft:` selector.

> 
> Second, while I'm still concerned with forcing the developer with override a NextFocusIntent with an InsertTabAction, I can't think of a better solution. However, this requires much more education, since overriding stuff is more confusing. We need a way to tell the developers to override the intent-to-action map, instead of the activator-to-intent map (which is understandably more intuitive), and the breakage otherwise will only be noticed in the corner case of emoji picker. Moreover, they will have to remember the enumerated intents to override, which basically means they're associating `NextFocusIntent` with pressing a `Tab` key.

Agreed. Ideally the framework should invoke whatever intent is mapped to Esc key above the editor. Users would expect that Shortcut with tab activator above text field will work, and it indeed will, except on macOS. Unfortunately I don't think this is possible to do currently with Shortcuts.
 
> I'm wondering if it will be a better choice to create a `MacosSelectorIntent(String content)` and map them to the actions for macOS. This also makes it look less weird to map to an `InsertTabAction`. What do you think?

Which selectors would this be for? If I understand correctly you suggesting mapping

`'insertTab:': MacosSelectorIntent('insertTab:),` and then having `NextFocusAction` registered for `MacosSelectorIntent('insertTab:`)? What about `cancelOperation:` selector? That is currently mapped to `DismissIntent`, if that gets changed to `MacosSelectorIntent('cancelOperation:')` it might be confusing.
> 
> And also, I definitely prefer documenting the data flow above somewhere in the engine.

**Comment by dkwingsmt** [comment]
We'll want to figure out how to best solve this problem after the fix. It's not out of the question to recommend overriding the intent-action table as the preferred way for _all_ platforms instead of the activator-intent table, but I think this will not work for the Web, since the Web rejects control events such as arrow keys with the activator-intent table once and for all, and there are no second calls with the intent-action table. @knopp Can you verify if this is true?

**Comment by dkwingsmt** [comment]
>

**Comment by dkwingsmt** [comment]
(Misclicked closing.)

**Comment by dkwingsmt** [comment]
> It rejects ArrowLeft, because it is mapped to `DirectionalFocusIntent(TraversalDirection.left)` inside `app.dart`. So during text editing it needs to be mapped to `DoNothingAndStopPropagationTextIntent` in order for framework to not handle the event.

I see. But `_defaultAppleOsShortcuts` also includes Meta+Arrow keys. Don't you need to reject them as well? Or it's just because they're not used in the emoji picker? If so, we probably want to document this under `_defaultAppleOsShortcuts`, something like:

> On macOS, some key combinations must not be handled by the app by default, so that they can be handled by the IME (mostly, the emoji picker) first, and sent back as selectors, converted to intents as a second round handling. These key combinations must either be excluded from `_defaultAppleOsShortcuts`, or overridden in `_macShortcuts` in `default_text_editing_shortcuts.dart`. If the following key combinations are added, please also add them to the text `_macShortcuts`.
>
> Alt-Arrow keys

**Comment by dkwingsmt** [comment]
Also cc @gspencergoog

**Comment by knopp** [comment]
> I see. But `_defaultAppleOsShortcuts` also includes Meta+Arrow keys. Don't you need to reject them as well? Or it's just because they're not used in the emoji picker? If so, we probably want to document this under `_defaultAppleOsShortcuts`, something like:
> 
My bad. I missed those. We need to override those too. It only worked because the `ScrollAction` checks if there are any clients of primaryScrollController, which there weren't, so the action was enabled.

**Comment by knopp** [comment]
> * We could recommend overriding intent-action table as the preferred way from today on, but I suspect this will not work on Web.

I'm not sure I understand your point correctly. Binding tab action to `NextFocusIntent` using the `Action` widget (instead of using Shortcut widget) works on all platforms including web.

**Review by LongCatIsLooong — APPROVED** [review]
Thank you for looking into this! LGTM.

I assume this has to be merge together with the framework pull request?

**Comment by chinmaygarde** [comment]
Is this blocked on merging? May I autosubmit?

**Comment by knopp** [comment]
Not yet, sorry. There's one outstanding issue I need to investigate in the framework PR.

**Comment by knopp** [comment]
@chinmaygarde, this should be good to go, after the tests pass. I needs to be merged together with the [framework PR](https://github.com/flutter/flutter/pull/105407), not sure how to do that.

**Comment by dkwingsmt** [comment]
I think merging the engine PR firsy shouldn't break anything. It adds the functionality that unhandled key presses will be converted to selectors, but since these text shortcuts are still handled, this functionality should not be triggered.

**Pull request merged**
